Add upstream cookie support

TCY16 opened the pull request:
- [x] Create .rpl test with required changed in testbound
- [x] Make cookies privacy aware (i.e. exclusive per source IP address)
    - [ ] Make test for privacy aware cookies
- [x] Make upstream cookies configurable

Commit:
add functionality to store the client cookie and attach both client and server cookie to the upstream query

Commit:
store complete cookie when receiving from downstream. Add function through search through an edns_option list

Commit:
create random cookies by creating and storing a random state in infra_cache

Commit:
Merge branch 'master' of github.com:NLnetLabs/unbound into features/upstream-cookies

Commit:
add verification of the client cookie in the response and add cookie in the rrdef enum

Commit:
add unit test for server cookie init

Commit:
add the 3 teststeps for upstream cookies

Commit:
make upstream cookies configurable

Commit:
fix memory alloc bug for the testcase and c89 style

Commit:
fix incorrect copy error

Commit:
add step in rpl to verify the updated server cookie and fix bug with unupdated rdlen

Commit:
fix test and add mapage entry

wcawijngaards commented on services/cache/infra.c:
```diff
@@ -383,12 +388,24 @@ static void
+
+	for (i = 0; i < 8; i++) {
+		cookie[i] = ub_random_max(infra->random_state,
+			255); // @TODO is 255 correct? macro-ify it?
```
It needs to be 256.
```suggestion
			256);
```

wcawijngaards commented on services/cache/infra.c:
```diff
@@ -685,6 +706,119 @@ infra_edns_update(struct infra_cache* infra, struct sockaddr_storage* addr,
+		lock_rw_unlock(&e->lock);
+	}
+
+	return &data->cookie;
```
Here data is returned that is not locked, but it points into the entry in the cache's data element. That is a race condition, if the entry->data->cookie is accessed or modified.

wcawijngaards commented on services/cache/infra.c:
```diff
@@ -685,6 +706,119 @@ infra_edns_update(struct infra_cache* infra, struct sockaddr_storage* addr,
+
+	/* the client cookie was set on the outgoing upstream, so the entry
+	 * must exists here */
+	assert(e);
```
This is not true, because once the reply comes in, the entry may have fallen out of the cache. By the cache getting full and the item dropping out of it, or potentially another thread has edited the contents. So the item could not be there any more. Insert it anew, or ignore it, or something like that.

wcawijngaards commented on services/cache/infra.h:
```diff
@@ -52,6 +52,42 @@
+/**
+ * The actual EDNS cookie data
+ */
+union edns_cookie_data {
```
This does not work, I think, the 24 byte is not the same length as the struct components underneath. The struct members get structure alignment. And the members of that struct are not guaranteed to be at the correct spots in the 24 byte sequence, I believe.

wcawijngaards commented on services/cache/infra.h:
```diff
@@ -318,6 +359,40 @@ int infra_edns_update(struct infra_cache* infra,
+ * @param name: name of zone
+ * @param namelen: length of name
+ * @param timenow: what time it is now.
+ * @return: struct ends_cookie*
```
```suggestion
 * @return: struct edns_cookie*
```
Spelling fixed.

wcawijngaards commented on testcode/testpkts.c:
```diff
@@ -1499,6 +1516,92 @@ match_ednsdata(uint8_t* q, size_t qlen, uint8_t* p, size_t plen)
+		verbose(3, "No edns opt, no cookie");
+		return 0;
+	}
+
```
You need an additional if-statement on walk_query_len < 12, and if so, it is too short for the read_uint16 you do further on.

wcawijngaards commented on testcode/testpkts.c:
```diff
@@ -1499,6 +1516,92 @@ match_ednsdata(uint8_t* q, size_t qlen, uint8_t* p, size_t plen)
+
+	walk_query += 8;
+	walk_query_len -= 8;
+
```
An additional if-statement, that walk_query_len < 4 means it is too short for the read_uint16 further on.

wcawijngaards commented on testcode/testpkts.c:
```diff
@@ -1499,6 +1516,92 @@ match_ednsdata(uint8_t* q, size_t qlen, uint8_t* p, size_t plen)
+		verbose(3, "EDNS cookie is not 24 bytes, so not a correct complete cookie");
+		return 0;
+	}
+
```
You need an additional if-statement that walk_query_len allows for that 24 length more data, and if shorter, return failure.

wcawijngaards commented on testcode/testpkts.c:
```diff
@@ -1690,6 +1803,120 @@ adjust_packet(struct entry* match, uint8_t** answer_pkt, size_t *answer_len,
+			if (sldns_read_uint16(walk_query+2) == 8) {
+				/* copy the EDNS client cookie from the query
+				 * packet to the response */
+				memcpy(walk_response, walk_query, 12);
```
memcpy needs aligned pointers, and these are not, they are packet offsets. memmove can copy for these kinds of possibly unaligned places, also for memcpy calls on walk_response that you write further on.

wcawijngaards commented on testcode/testpkts.c:
```diff
@@ -1690,6 +1803,120 @@ adjust_packet(struct entry* match, uint8_t** answer_pkt, size_t *answer_len,
+				sldns_write_uint16(rdlen_ptr_response, 28);
+				sldns_write_uint16(walk_response+2, 24);
+
+				reslen = origlen + 28;
```
reslen was already incremented by 28 at line 1818, I do not think you have to do it again.

wcawijngaards commented on testcode/unitmain.c:
```diff
@@ -469,11 +469,12 @@ infra_test(void)
 	struct infra_key* k;
 	struct infra_data* d;
 	int init = 376;
+	struct ub_randstate* rnd = ub_initstate(NULL);
```
Put at line 531, at the end of this function, `ub_randfree(rnd);` to delete it.

wcawijngaards reviewed (commented):
So I noticed some debug and todo comments and debug log prints, and made some suggestions in the items. The code looks nice. The .rpl file does not exercise the outside_network code, but it is not that big a part of the code.

Commit:
implement @wcawijngaards' review comments

Commit:
Merge branch 'master' into features/upstream-cookies

Commit:
add privacy aware cookies. This breaks the rpl test

wtoorop committed:
extended_error_encode() for extended errors

wtoorop committed:
Downstream DNS Cookies a la RFC7873 and RFC9018

Create server cookies for clients that send client cookies.
Needs to be turned on in the config file with:

	answer-cookie: yes

A cookie-secret can be configured for anycast setups.
Also adds an access control list that will allow queries with
either a valid cookie or over a stateful transport.

wtoorop committed:
The generated lexer and parser sources for configuring cookies

Commit:
intermediate commit

wtoorop committed:
Merge branch 'master' into devel/merge-master-into-downstream-cookies

Commit:
intermediate commit v2; now with seemingly working private cookies

Commit:
add handover text and tests

Commit:
intermediate commit V3; Working up to test
